src/main/java/org/cloudfoundry/promregator/auth/OAuth2XSUAAEnricher.java

```diff
@@ -1,188 +1,86 @@
-		 * We won't be calling often anyway, so the server would be drained from resources.
-		 * Moreover, if the server has gone away in the meantime, the next attempt to 
-		 * call would fail with a recv error when reading from the socket.
-		 */
-		httpPost.setHeader("Connection", "close");
-		
-		List<NameValuePair> form = new ArrayList<>();
-		form.add(new BasicNameValuePair("response_type", "token"));
-		UrlEncodedFormEntity formEntity = new UrlEncodedFormEntity(form, Consts.UTF_8);
-		httpPost.setEntity(formEntity);
-		
-		CloseableHttpResponse response = null;
-		String json = null;
+	private final String getJWTAndCatchExceptions() {
+		String jwt = null;
 		try {
-			try {
-				response = httpclient.execute(httpPost);
-			} catch (ClientProtocolException e) {
-				log.error("Unable to read from the token server", e);
-				return null;
-			} catch (IOException e) {
-				log.error("IO Exception while reading from the token server", e);
-				return null;
-			}
-			
-			if (response.getStatusLine().getStatusCode() != 200) {
-				log.error(String.format("Server did not respond with ok while fetching JWT from token server; status code provided: %d", response.getStatusLine().getStatusCode()));
-				return null;
-			}
-			
-			try {
-				json = EntityUtils.toString(response.getEntity(), StandardCharsets.UTF_8);
-			} catch (ParseException e) {
-				log.error("GSON parser exception on JWT response from token server", e);
-				return null;
-			} catch (IOException e) {
-				log.error("IO Exception while running GSON parser", e);
-				return null;
-			}
-		} finally {
-			if (response != null) {
-				try {
-					response.close();
-				} catch (IOException e) {
-					log.info("Unable to properly close JWT-retrieval connection", e);
-				}
-			}
-		}
-		
-		if (json == null) {
-			log.warn("Null-JSON detected on OAuth response");
-			return null;
+			jwt = getJWT();
+		} catch (TokenFlowException | RuntimeException e) {
+			log.error("Unable to enrich request with JWT", e);
 		}
-		
-		Gson gson = new Gson();
-		TokenResponse oAuthResponse = gson.fromJson(JsonSanitizer.sanitize(json), TokenResponse.class);
-		
-		String jwt = oAuthResponse.getAccessToken();
-		log.info(String.format("JWT token retrieved: %s...", jwt.substring(0, Math.min(jwt.length() / 2, 30))));
-
-		int timeOutForUs = Math.max(oAuthResponse.getExpiresIn() - 30, oAuthResponse.getExpiresIn() / 2);
-		this.validUntil = Instant.now().plus(timeOutForUs, ChronoUnit.SECONDS);
-		
-		log.info(String.format("JWT is valid until %s", this.validUntil.toString()));
-		
 		return jwt;
 	}
 
-	private static class TokenResponse {
-		private String access_token;
-		private int expires_in;
-		
-		public String getAccessToken() {
-			return access_token;
-		}
+	private final String getJWT() throws TokenFlowException {
+		return this.tokenClient.execute().getAccessToken();
+	}
 
-		public int getExpiresIn() {
-			return expires_in;
+	public void close() throws IOException {
```

Comment:
Not sure where to call this `close()` method.

Reply:
Good point. So far, Promregator is not intended for a shutdown - instead we restart the entire application. Hence there is also no closing mechanism available yet. 
In fact this is already a deficit of [`AuthenticationEnricher`](https://github.com/promregator/promregator/blob/d67a182c4fc7fbfd989cde1eeb0b662196637e5c/src/main/java/org/cloudfoundry/promregator/auth/AuthenticationEnricher.java).

So, providing one right now is the best thing we can do right now.

Reply:
> In case the whole application is not designed for a proper shutdown anyway: Would it be possible not so solve that as part of this PR in order to keep that PR focused.

Yes, that is the right thing. 
However, I would feel glad to see, if we documented our concerns with a small `// TODO` comment here, "blaming" it to the missing method in the interface ("Ross und Reiter nennen").

BTW: We should consider implementing `Closable` here (and potentially already at `AuthenticationEnrichers`).

Reply:
`OAuth2XSUAAEnricher` implements now `Closeable`. We have now a `TODO` comment on the`close()` method.
Of course we can push `Closable` upwards to `AuthenticationEnricher` but the other two Enrichers does not have any resources which must be closed. For reasons of a symmetric handling that might make sense nevertheless ...

Reply:
I agree to stick with the current state.
I fear that we will open up a can of worms, if we pushed `Closable` upwards to `AuthenticationEnricher` now. It feels like that this change should go into a separate PR.